### PR TITLE
[Tooling] Fix version names and codes per variant

### LINKED ISSUE #14886: Improve flavor and version management in build.gradle
With the introduction of the new properties file for version management, added complexity was introduced to our build.gradle. We should look into improving how we're managing our product flavors and the different versions associated with them. 

PR: https://github.com/wordpress-mobile/WordPress-Android/pull/14613

Small things that should also be addressed in the same breadth include improving clarity around any properties needed for gradle tasks: 
1. https://github.com/wordpress-mobile/WordPress-Android/pull/14613#discussion_r655666569

### PR DESCRIPTION
**Pull request by AliSoftware**
This fixes an issue – introduced in https://github.com/wordpress-mobile/WordPress-Android/pull/14613 when we worked on adding support for the Jetpack app to our `build.gradle` and tooling – with the project versioning configuration.

The way the project configuration was modified in https://github.com/wordpress-mobile/WordPress-Android/pull/14613 (to introduce `version.properties`) made the per-variant resolution of `versionName` and `versionCode` incorrect.

Closes #14886

#### Test `printVersionName` & `printAllVersions`

 - `./gradlew -q printVersionName -PproductKey="wordpress" ` should return the value of `wordpress.versionName` in `version.properties`, aka `17.6-rc-4`
 - `./gradlew -q printVersionName -PproductKey="wordpress" -PbuildType="zalpha"` should return the value of `wordpress.zalpha.versionName` in `version.properties`, aka `alpha-301`
 - `./gradlew -q printVersionName -PproductKey="jetpack" ` should return the value of `jetpack.versionName` in `version.properties`, aka `17.6-rc-3`
 - `./gradlew -q printVersionName -PproductKey="jetpack" -PbuildType="zalpha"` should fall back to returning the value of `jetpack.versionName` in `version.properties`, aka `17.6-rc-3` – given there is no zalpha-specific versionName defined in `version.properties` for Jetpack
 - `./gradlew -q printAllVersions` should print the expected versionName and versionCode values for each variant

#### Test the version bumps

 - Cut a branch from this one, and name it with `release` in its name, eg. `test/release/version-fix` (because pre_checks look for 'release' in the current branch name)
 - Run `bundle exec fastlane run android_bump_version_beta app:wordpress`. It should fail when trying to commit/push, but should have updated the values in `version.properties` for **both** `wordpress.version{Name,Code}` and `wordpress.zalpha.version{Name,Code}`.
 - Run `bundle exec fastlane run android_bump_version_beta app:jetpack`. It should fail at the step it tries to commit/push, but should have updated the values in `version.properties` for `jetpack.version{Name,Code}` (no alpha for jetpack).

#### Test that an actual alpha build uses the alpha versions

 - Run `./gradlew bundleWordpressZalphaRelease` then open the generated `aab` in Android Studio (`open -a "Android Studio" WordPress/build/outputs/bundle/wordpressZalphaRelease/org.wordpress.android-wordpress-zalpha-release.aab`) and check that its `versionName` is `alpha-301` and its `versionCode` is `1066`

### Regression Notes

1. Potential unintended areas of impact

Wrong versioning used when building various flavors in various contexts (CI: Installable Builds, Beta and Release Builds)

2. What I did to test those areas of impact (or what existing automated tests I relied on)

 - Ran the tests above
 - Waiting to check the status on [the release build I triggered on CI from this branch](https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-Android/24953/workflows/c61d93fc-e73d-423c-b8e0-ee15378de355) to check if it uploads the expected versionNames and versionCodes to the expected tracks on PlayStore

3. What automated tests I added (or what prevented me from doing so)

N/A

### PR submission checklist:

- [x] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Gemfile.lock
Comment (AliSoftware) [inline Gemfile.lock:3]: This is pending a release in https://github.com/wordpress-mobile/release-toolkit/pull/283

I'm ok to merge this pointing to a temp branch now and switch back to pointing at `1.3.1` once the toolkit PR get approved and merged later and `1.3.1` gets released.

File: version.properties
Comment (AliSoftware) [inline version.properties:8]: This relies on the fact that the pending changes in https://github.com/wordpress-mobile/release-toolkit/pull/283 that this PR points to (see `fastlane/PluginFile` and `Gemfile.lock`) now know that the name of the flavor for alpha builds is `zalpha`, see [here](https://github.com/wordpress-mobile/release-toolkit/pull/283/files#diff-6f51a72fba2128ef36317c1bf3b66ea1f8c47b7f38f2306f9c4b9b34bdfd618aR100-R108)

We intend to improve that to clean up the code in release toolkit later (especially not hardcode that value in release toolkit – which was hardcoded as `alpha.` before and is now `zalpha` – nor rely on the `HAS_ALPHA_VERSION` env var, but instead pass the alpha flavor name as parameter of each action) but this clean up of the toolkit will be for another day – tracked in https://github.com/wordpress-mobile/release-toolkit/issues/277

File: WordPress/build.gradle
Comment (oguzkocer) [inline WordPress/build.gradle:538]: I think it'd be better if we get an error when the `productKey` is not passed in just so we don't forget it.

I'd go one step further and make this an internal plugin and use a named command line argument if we want to keep this as is. (I imagine we might want something more robust as a central plugin that we can use across the apps 🤔 )

File: WordPress/build.gradle
Comment (oguzkocer) [inline WordPress/build.gradle:548]: I wonder if we should iterate through all the outputs just to cover future changes 🤔 

Maybe we can do something like this:

```suggestion
            def apkData = variant.outputs.collect { it.apkData }
            println "${variant.name}: ${apkData.collect { it.versionName }} ${apkData.collect { it.versionCode }}"
```

This will put the versionName and versionCode between `[]` since they'll be lists, but that probably doesn't matter too much since this is essentially a debug task.

What do you think @AliSoftware ?

**Review by oguzkocer — APPROVED** [review]
@AliSoftware This is great, thank you for taking care of it! I've left a couple minor suggestions, but nothing that'd hold off this review.

Aside from those, the one thing that bothers me a bit is that we are using fallback in the version names and using actual values from other variants for it. I think that can get confusing for us. For instance, a couple days ago there was a conversation about using these values to figure out `debug` vs `release` in Tracks.

What if we take this opportunity to add some suffixes for these fallbacks, so it's obvious what they are related to. For example, for every `debug` build, we can add a `-debug` suffix. For missing variants, we can add `-missing-{flavorName}` - well hopefully something better, but you get the idea.

I understand these versions are not important for Play Store, but they are still used for other tasks and we should assume `BuildConfig.VERSION_NAME` is used for something important within the project (actual source code) and configure it accordingly.

Let me know what you think!

File: WordPress/build.gradle
Comment (AliSoftware) [reply WordPress/build.gradle:538]: +1 for error if productKey not passed.

As for making this an internal plugin… not sure it's worth the effort to set up a whole plugin just for that? 🤔

File: WordPress/build.gradle
Comment (AliSoftware) [reply WordPress/build.gradle:548]: Works for me. I actually discovered a nice syntax while browsing for Groovy syntax for `.map` / `.collect` yesterday, it seems `*.` is a shorthand for map/collect. Just tried it and seems to work well, while keeping the code compact and really nice to read:

```suggestion
            def apkData = variant.outputs*.apkData
            println "${variant.name}: ${apkData*.versionName} (${apkData*.versionCode})"
```

```
$ ./gradlew -q printAllVersions

Checking whether `configure` binary is present
wordpressWasabiDebug: [17.6-rc-4] ([1065])
wordpressZalphaDebug: [alpha-301] ([1066])
wordpressJalapenoDebug: [17.6-rc-4] ([1065])
wordpressVanillaDebug: [17.6-rc-4] ([1065])
jetpackWasabiDebug: [17.6-rc-3] ([1063])
jetpackZalphaDebug: [17.6-rc-3] ([1063])
jetpackJalapenoDebug: [17.6-rc-3] ([1063])
jetpackVanillaDebug: [17.6-rc-3] ([1063])
wordpressWasabiRelease: [17.6-rc-4] ([1065])
wordpressZalphaRelease: [alpha-301] ([1066])
wordpressJalapenoRelease: [17.6-rc-4] ([1065])
wordpressVanillaRelease: [17.6-rc-4] ([1065])
jetpackWasabiRelease: [17.6-rc-3] ([1063])
jetpackZalphaRelease: [17.6-rc-3] ([1063])
jetpackJalapenoRelease: [17.6-rc-3] ([1063])
jetpackVanillaRelease: [17.6-rc-3] ([1063])
```

**Comment by AliSoftware** [comment]
> Aside from those, the one thing that bothers me a bit is that we are using fallback in the version names and using actual values from other variants for it. I think that can get confusing for us. For instance, a couple days ago there was a conversation about using these values to figure out `debug` vs `release` in Tracks.
> 
> What if we take this opportunity to add some suffixes for these fallbacks, so it's obvious what they are related to. For example, for every `debug` build, we can add a `-debug` suffix. For missing variants, we can add `-missing-{flavorName}` - well hopefully something better, but you get the idea.

I get the concern. The tricky part tho is that:
 - Our process only bumps one of the key for release builds. For example when we call `android_bump_beta_version` or `android_bump_final_release_version`, it will modify the `wordpress.versionName` and `wordpress.versionCode`.
 - If instead we put a rule which adds a suffix (my suggestion would be for `+{flavorName}` suffix, e.g. `17.6-rc-4+jalapeno` – that's the way semver.org does metadata in version names iinm), that means that to avoid having `17.6+vanilla` for release builds, and instead just use `17.6` without suffix, we would need to have an explicit `wordpress.vanilla.versionName` (instead of just `wordpress.versionName` and relying on fallback)
 - But then that would mean that `jalapeno` and `wasabi` would not have any entry to fall back to anymore, except if we _also_ keep `wordpress.versionName` in the `version.properties`. But that means we would have to keep both `wordpress.vanilla.versionName` (for release builds) and `wordpress.versionName` (the fallback for others) in sync, always bumping them in concert. Which is something that `android_bump_beta_version` and `android_bump_release_version` don't handle

So instead, I think the better way would be to keep the `version.properties` as is, and handle the decision for suffix or not at the `build.gradle` level. For example, hardcode a condition in `versionForVariant` that `if (buildType != 'vanilla' && didNotFindExplicitEntryButDidFallbackInstead) { versionName += "+${buildType}"` or something. wdyt?

File: WordPress/build.gradle
Comment (AliSoftware) [reply WordPress/build.gradle:548]: Actually directly pushed this change on `release/17.6` just now to not forget about it (was trivial enough that I figured it was not worth a PR): 441604e104

**Comment by AliSoftware** [comment]
@oguzkocer Another/alternative solution to the one above: don't use suffix in `versionName` to differentiate the flavors, to avoid over-complicating the logic (given the concern I explained above comment)… and instead use either `applicationIdSuffix` (which, in fact, we already have for `jalapeno` and `wasabi`), or maybe even better, use `buildConfigField` – which seems more idiomatic to me for those kind of needs? – to provide that information as a dedicated value exposed to Java/Kotlin.

File: WordPress/build.gradle
Comment (oguzkocer) [reply WordPress/build.gradle:538]: It's probably easier than you expect, but I don't think this should be the first thing we move to internal plugins. I hope to work on this as part of the modularization project, and if I do, I'll just make this a part of it.

File: WordPress/build.gradle
Comment (oguzkocer) [reply WordPress/build.gradle:548]: Fwiw, I am not a big fan of merging any change, however trivial it might be, without giving a chance for the reviewer to check it out. I understand why you preferred to do it this way and I've also done this on a few occasions, so all good 👍

**Comment by oguzkocer** [comment]
@AliSoftware I was thinking of doing the suffix at the `build.gradle` level, when we define each variant or calculate its version. And my preference would be to have the suffix as part of the versionName because it's just easier to work with. That way, we have to tackle the issue just once (when defining version names) and we can put unit tests to verify the behavior. If we use an extra property, we have to remember to use that everywhere it's necessary.

I don't think either of us have the bandwidth to work on this right now, so let's just leave it as is for now and let's maybe see if we can find a more generic approach we can use across other client apps when we do tackle it. Does that work for you?

**Comment by AliSoftware** [comment]
Yep makes sense.

I'll likely have to re-work on this soon when I try to undo the per-app separation of versioning and to go back to a way where WP's and JP's versionName and versionCode are always in sync and built by CI in concert during the same workflow – which is how iOS does it, and how we should have done it in WPAndroid in the first, place, see brainstorm/discussion in private Slack channel. So we might end up removing the app param and the `jetpack.*` and `wordpress.*` prefixes in the `version.properties` and the per-variant / per-app logic we have put in place to do something simpler anyway…

**Pull request merged**
